components/channel_members_rhs/index.ts

```diff
@@ -33,6 +34,57 @@ import {loadMyChannelMemberAndRole} from 'mattermost-redux/actions/channels';
 
 import RHS, {Props, ChannelMember} from './channel_members_rhs';
 
+const getProfiles = createSelector(
+    'getProfiles',
+    getProfilesInCurrentChannel,
+    getUserStatuses,
+    getTeammateNameDisplaySetting,
+    getMembersInCurrentChannel,
+    (profilesInCurrentChannel, userStatuses, teammateNameDisplaySetting, membersInCurrentChannel) => {
+        const channelAdmins: ChannelMember[] = [];
+        const channelMembers: ChannelMember[] = [];
+        profilesInCurrentChannel.forEach((profile) => {
+            const member = {
+                user: profile,
+                membership: membersInCurrentChannel[profile.id],
+                status: userStatuses[profile.id],
+                displayName: displayUsername(profile, teammateNameDisplaySetting),
+            } as ChannelMember;
+
+            if (member.membership) {
+                // group by role unless we are doing a search
+                if (isChannelAdmin(member.membership)) {
+                    channelAdmins.push(member);
+                    return;
+                }
+
+                channelMembers.push(member);
+            }
+        });
+        return [channelAdmins, channelMembers];
+    },
+);
+
+const searchProfiles = createSelector(
+    'searchProfiles',
+    (state: GlobalState, search: string) => searchProfilesInCurrentChannel(state, search, false),
```

Comment:
Note that since the selector is taking an argument, it's possible that two copies of a component calling this with different arguments would overwrite each other's memoized results. Generally, we'd want to use a factory or something to prevent that, but since this component is defined alongside a component that's essentially a singleton, I don't think anything needs to be changed here

Reply:
I'm gonna keep it as it is but just to make sure, you meant something like this?
```ts
const makeSearchProfiles(searchTerm: string) {
	  return createSelector(
	      'searchProfiles',
	      (state: GlobalState) => searchProfilesInCurrentChannel(state, searchTerm, false),
	      getUserStatuses,
	      getTeammateNameDisplaySetting,
	      getMembersInCurrentChannel,
	      (profilesInCurrentChannel, userStatuses, teammateNameDisplaySetting, membersInCurrentChannel) => {
	          const channelMembers: ChannelMember[] = [];
	          profilesInCurrentChannel.forEach((profile) => {
	              channelMembers.push({
	                  user: profile,
	                  membership: membersInCurrentChannel[profile.id],
	                  status: userStatuses[profile.id],
	                  displayName: displayUsername(profile, teammateNameDisplaySetting),
	              });
	          });
	          return [[] as ChannelMember[], channelMembers];
	      },
	  );
}
```

and usage

```ts
makeSearchProfiles('Harrison')(state);
```

Reply:
Almost. You'd still pass the `searchTerm` in when actually calling the selector, but you'd generate a new copy of `searchProfiles` for each instance of the component. You can do that by while essentially giving each instance of the component its own `mapStateToProps` by calling `connect` slightly differently.

```javascript
function makeSearchProfiles() {
    return /* The exact createSelector statement you have currently in the PR */;
}

function makeMapStateToProps() {
    // makeMapStateToProps is called for each instance of the ChannelMembersRHS, so each one has its own
    // copy of searchProfiles which is memoized independently
    const searchProfiles = makeSearchProfiles();
    
    return (state, ownProps) => {
        return {
            profiles: searchProfiles(state, ownProps.searchTerms),
        };
    };
}

return connect(makeMapStateToProps)(ChannelMembersRHS);
```

You can find examples of that